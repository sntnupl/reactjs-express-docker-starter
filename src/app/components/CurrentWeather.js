import React from 'react';
import PropTypes from 'prop-types';
import {weatherCity} from '../services/WeatherService';

const WeatherSummary = ({data, error}) => {
    if (!data) return null;
    if (error && error.msg) return null;

    const {name, desc, curr, min, max} = data;
    if (!name) return null;

    return (
        <div className="weather-data-card">
            <p>{name}</p>
            <ul>
                <li>Current Weather: {desc}</li>
                <li>Current Temp: {curr} deg celcius</li>
                <li>Min Temp: {min} deg celcius</li>
                <li>Max Temp: {max} deg celcius</li>
            </ul>
        </div>
    );
};

WeatherSummary.propTypes = {
    data: PropTypes.shape({
        name: PropTypes.string,
        desc: PropTypes.string,
        curr: PropTypes.number,
        min: PropTypes.number,
        max: PropTypes.number,
    }),
    error: PropTypes.shape({
        msg: PropTypes.string,
    }),
};




class CurrentWeather extends React.Component {
    static Summary = WeatherSummary;
    static defaultProps = {
        onData: () => {}
    };

    constructor(props) {
        super(props);
    }
    refLocation = null;
    state = {
        error: {
            msg: undefined
        },
        isFetchingData: false,
        weatherData: {
            name: undefined,
            desc: undefined,
            curr: undefined,
            min: undefined,
            max: undefined
        },
    };

    handleWeatherData = (resp) => {
        const {main, weather, name} = resp.data;
        this.setState(() => {
            return {
                isFetchingData: false,
                weatherData: {
                    name,
                    desc: weather[0].description,
                    curr: main.temp,
                    min: main.temp_min,
                    max: main.temp_max
                },
            };
        });

        this.props.onData(resp);
    };

    onWeatherFetchError = (err) => {
        console.error('CurrentData error: ', JSON.stringify(err));
        this.setState(() => {
            return {
                error: {msg: 'api call failed'},
                isFetchingData: false
            };
        });
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.refLocation.value) return;

        this.setState(() => {
            return {
                error: {msg: undefined},
                isFetchingData: true,
            };
        });

        const location = this.refLocation.value;
        weatherCity(location.trim())
            .then(resp => this.handleWeatherData(resp))
            .catch(err => this.onWeatherFetchError(err));
        this.refLocation.value = '';
    };

    render = () => {
        const {error, isFetchingData, weatherData} = this.state;
        const children = React.Children.map(this.props.children, (child) => {
            return React.cloneElement(child, {
                error,
                data: weatherData,
            });
        });

        return (
            <div className="current-weather">
                <div className="current-weather-form">
                    <form onSubmit={this.onSubmit}>
                        <input
                            type="text"
                            ref={(el) => {this.refLocation = el;}}
                            placeholder="city name (, country name)"/>
                        {!isFetchingData
                            ? <button type="submit">Fetch Weather Data</button>
                            : <button type='submit' disabled>Fetching..</button>
                        }
                        <span>{error.msg}</span>
                    </form>
                </div>
                {children}
            </div>
        );
    }
}

CurrentWeather.propTypes = {
    onData: PropTypes.func,
    children: PropTypes.node.isRequired
};

export default CurrentWeather;