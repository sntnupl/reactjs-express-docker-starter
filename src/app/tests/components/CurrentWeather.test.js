import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CurrentWeather from '../../components/CurrentWeather';

test('should render correctly', () => {
    const wrapper = shallow(
        <CurrentWeather>
            <CurrentWeather.Summary/>
        </CurrentWeather>
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
});


/* TODO List
*   1. unless form is submitted, the initial state has no weather data [D]
*   2. isFetchingData [D]
*   3. error.msg is displayed when available [D]
*   4. On Submit, weather data is fetched properly and state update is done properly
*   5. If Weather-Summary has no city name, it does not render summary [D]
*   6. If Weather-Summary has error, it does not render summary [D]
**/

test('no weather data is displayed on initial render', () => {
    const wrapper = shallow(
        <CurrentWeather>
            <CurrentWeather.Summary/>
        </CurrentWeather>
    );
    expect(wrapper.state().isFetchingData).toBe(false);
    expect(wrapper.state().weatherData.name).toBeFalsy();
    expect(wrapper.state().weatherData.curr).toBeFalsy();
});


test('submit button stays disabled until a weather data fetch is complete', () => {
    const wrapper = shallow(
        <CurrentWeather>
            <CurrentWeather.Summary/>
        </CurrentWeather>
    );
    expect(wrapper.find('button [disabled]').length).toBe(0);
    wrapper.setState({ isFetchingData: true });
    expect(wrapper.find('button [disabled]').length).toBe(1);
});


test('weather form shows error, if any', () => {
    const wrapper = shallow(
        <CurrentWeather>
            <CurrentWeather.Summary/>
        </CurrentWeather>
    );
    wrapper.setState({
        error: { msg: 'test error' }
    });
    const formNode = wrapper.find('form').children();
    expect(formNode.contains(<span>test error</span>));
});


test('no summary data is rendered if location name is unavailable', () => {
    const wrapper = shallow(<CurrentWeather.Summary />);
    wrapper.setProps({ data: { name: undefined } });
    expect(wrapper.find('.weather-data-card').length).toBe(0);
});


test('no summary data is rendered if there is error', () => {
    const wrapper = shallow(<CurrentWeather.Summary />);
    wrapper.setProps({ error: { msg: undefined } });
    expect(wrapper.find('.weather-data-card').length).toBe(0);
});


test('weather data is handled correctly', () => {
    const wrapper = shallow(
        <CurrentWeather>
            <CurrentWeather.Summary/>
        </CurrentWeather>
    );
    wrapper.setState({
        isFetchingData: true,
        weatherData: {}
    });
    wrapper.instance().handleWeatherData({
        data: {
            name: 'kolkata',
            main: {
                temp: 15,
                temp_min: 12,
                temp_max: 25,
            },
            weather: [{ description: 'test description' }],
        }
    });

    expect(wrapper.state().isFetchingData).toBe(false);
    expect(wrapper.state().weatherData.name).toEqual('kolkata');
    expect(wrapper.state().weatherData.curr).toEqual(15);
});

test('weather fetch error is handled correctly', () => {
    const wrapper = shallow(
        <CurrentWeather>
            <CurrentWeather.Summary/>
        </CurrentWeather>
    );
    wrapper.setState({
        isFetchingData: true,
        weatherData: {},
        error: {},
    });
    wrapper.instance().onWeatherFetchError({});

    expect(wrapper.state().isFetchingData).toBe(false);
    expect(wrapper.state().error.msg).toBeTruthy();
});

test('weather summary is displayed correctly', () => {
    const wrapper = shallow(<CurrentWeather.Summary/>);
    wrapper.setProps({
        data: {
            name: 'kolkata',
            desc: 'test desc',
            curr: 15,
            min: 12,
            max: 25,
        },
    });
    expect(wrapper.find('.weather-data-card').length).toEqual(1);
});
