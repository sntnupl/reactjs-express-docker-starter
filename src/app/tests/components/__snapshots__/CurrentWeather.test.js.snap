// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="current-weather-container"
>
  <div
    className="current-weather-form"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        placeholder="city name (, country name)"
        size="60"
        type="text"
      />
      <button
        type="submit"
      >
        Fetch Weather Data
      </button>
      <span />
    </form>
  </div>
  <WeatherSummary
    data={
      Object {
        "curr": undefined,
        "desc": undefined,
        "max": undefined,
        "min": undefined,
        "name": undefined,
      }
    }
    error={
      Object {
        "msg": undefined,
      }
    }
    key=".0"
  />
</div>
`;
